modules/ppcp-wc-gateway/src/Gateway/PayUponInvoice/PaymentSourceFactory.php

```diff
@@ -0,0 +1,59 @@
+<?php
+/**
+ * PUI payment source factory.
+ *
+ * @package WooCommerce\PayPalCommerce\WcGateway\Gateway\PayUponInvoice
+ */
+
+declare(strict_types=1);
+
+namespace WooCommerce\PayPalCommerce\WcGateway\Gateway\PayUponInvoice;
+
+use WC_Order;
+
+/**
+ * Class PaymentSourceFactory.
+ */
+class PaymentSourceFactory {
+
+	/**
+	 * Create a PUI payment source from a WC order.
+	 *
+	 * @param WC_Order $order The WC order.
+	 * @return PaymentSource
+	 */
+	public function from_wc_order( WC_Order $order ) {
+		$address    = $order->get_address();
+		$birth_date = filter_input( INPUT_POST, 'billing_birth_date', FILTER_SANITIZE_STRING ) ?? '';
```

Comment:
Retrieving HTTP input in this function feels a bit weird. Maybe better e.g. pass this value as a parameter.

It's usually a good idea to isolate and keep on higher levels I/O things like that.